pkg/disk/v1alpha2/disk.go

```diff
@@ -0,0 +1,251 @@
+	}
+}
+
+// IsSparse returns true if the spc is of sparse type.
+func IsSparse() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == string(apis.TypeSparseCPV)
+	}
+}
+
+// IsDiskActive returns true if the disk is active.
+func IsDiskActive() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.Status.State == DiskActiveState
+	}
+}
+
+// IsUsableDisk returns true if this disk can be used for pool provisioning.
+func IsUsableDisk(usedDiskMap map[string]int) Predicate {
+	return func(d *Disk) bool {
+		return usedDiskMap[d.Object.Name] == 0
+	}
+}
+
+// IsUsableNodeDisk returns true if disk of this node can be used for pool provisioning.
+func IsUsableNodeDisk(usedNodeMap map[string]bool) Predicate {
+	return func(d *Disk) bool {
+		return !usedNodeMap[d.GetNodeName()]
+	}
+}
+
+// IsDiskBelongToNode returns true if the disk belongs to the provided node.
+func IsDiskBelongToNode(nodeName string) Predicate {
+	return func(d *Disk) bool {
+		return d.GetNodeName() == nodeName
+	}
+}
+
+// IsType returns true if the disk is of type same as passed argument
+func IsType(diskType string) Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == diskType
+	}
+}
+
+// IsValidPoolTopology returns true if the topology is valid.
+func IsValidPoolTopology(poolType string, diskCount int) bool {
+	return DefaultDiskCount[poolType]%diskCount == 0
+}
+
+// IsDiskType returns true if the spc is of disk type.
+func IsDiskType() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == string(apis.TypeDiskCPV)
+	}
+}
+
+// GetNodeName returns the node name to which the disk is attached
+func (d *Disk) GetNodeName() string {
+	return d.Object.GetLabels()[string(apis.HostNameCPK)]
+}
+
+// Filter will filter the disk instances if all the predicates succeed against that spc.
+func (l *DiskList) Filter(p ...Predicate) *DiskList {
+	var plist predicateList
+	plist = append(plist, p...)
+	if len(plist) == 0 {
+		return l
+	}
+
+	filtered := NewListBuilder().List()
+	for _, spcAPI := range l.ObjectList.Items {
+		spcAPI := spcAPI // pin it
+		Disk := BuilderForAPIObject(&spcAPI).Disk
+		if plist.all(Disk) {
+			filtered.ObjectList.Items = append(filtered.ObjectList.Items, *Disk.Object)
+		}
+	}
+	return filtered
+}
+
+// NewBuilder returns an empty instance of the Builder object.
+func NewBuilder() *Builder {
+	return &Builder{
+		Disk: &Disk{&apis.Disk{}},
+	}
+}
+
+// BuilderForObject returns an instance of the Builder object based on spc object
+func BuilderForObject(Disk *Disk) *Builder {
+	return &Builder{
+		Disk: Disk,
+	}
+}
+
+// BuilderForAPIObject returns an instance of the Builder object based on spc api object.
+func BuilderForAPIObject(spc *apis.Disk) *Builder {
+	return &Builder{
+		Disk: &Disk{spc},
+	}
+}
+
+// GetDeviceID returns the device ID of the disk.
+func (d *Disk) GetDeviceID() string {
```

Comment:
Can we explain in code comments what exactly is device id. Why can it be device link or a path?

Reply:
done